hedera-node/hedera-networkadmin-service-impl/src/main/java/com/hedera/node/app/service/networkadmin/impl/handlers/FreezeHandler.java

```diff
@@ -54,77 +61,136 @@ public FreezeHandler() {
         // Dagger2
     }
 
+    /**
+     * This method is called during the pre-handle workflow for Freeze transactions.
+     *
+     * @param context the {@link PreHandleContext} which collects all information that will be
+     *     passed to {@code handle()}
+     * @see <a href="https://hashgraph.github.io/hedera-protobufs/#freeze.proto">Protobuf freeze documentation</a>
+     */
     @Override
-    @SuppressWarnings("java:S1874") // disable the warnings for use of deprecated code
-    // it is necessary to check getStartHour, getStartMin, getEndHour, getEndMin, all of which are deprecated
-    // because if any are present then we set a status of INVALID_FREEZE_TRANSACTION_BODY
     public void preHandle(@NonNull final PreHandleContext context) throws PreCheckException {
         requireNonNull(context);
+        pureChecks(context.body());
+
+        // from proto specs, it looks like update file not required for FREEZE_UPGRADE and TELEMETRY_UPGRADE
+        // but specs aren't very clear
+        // current code in FreezeTransitionLogic checks for the file in specialFiles
+        // so we will do the same
+        final FreezeTransactionBody freezeTxn = context.body().freezeOrThrow();
+        final FreezeType freezeType = freezeTxn.freezeType();
+        if (Arrays.asList(FREEZE_ONLY, FREEZE_UPGRADE, TELEMETRY_UPGRADE).contains(freezeType)) {
+            final Timestamp txValidStart = context.body().transactionIDOrThrow().transactionValidStartOrThrow();
+            verifyFreezeStartTimeIsInFuture(freezeTxn, txValidStart);
+        }
+        if (Arrays.asList(FREEZE_UPGRADE, TELEMETRY_UPGRADE, PREPARE_UPGRADE).contains(freezeType)) {
+            final ReadableSpecialFileStore specialFileStore = context.createStore(ReadableSpecialFileStore.class);
+            verifyUpdateFileAndHash(freezeTxn, specialFileStore);
+        }
 
-        FreezeTransactionBody freezeTxn = context.body().freeze();
+        // no need to add any keys to the context because this transaction does not require any signatures
```

Comment:
I think we should still add required keys if any are known to exist.  The privileged account service **may** allow overriding requirements, but that should not change what is required.
If we know, for instance, that this transaction requires a specific account, then we should probably add that here as an additional layer of security.

Reply:
Freeze service really doesn't require any signatures.